### PR TITLE
Move SymDispatchMode to its own file

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112038
* #112037
* #112036
* __->__ #112035

This is just code movement + a getter and a setter to break the
dependency of SymDispatchMode, and in turn, ProxySymDispatchMode on
sympy.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -144,16 +110,15 @@ def create_contiguous(shape):
     return list(reversed(strides))
 
 def _handle_sym_dispatch(func, args, kwargs):
```
Comment (peterbell10) [inline torch/fx/experimental/symbolic_shapes.py:112]: Any reason not to put this in `_sym_dispatch_mode.py`?

Comment (lezcano) [reply torch/fx/experimental/symbolic_shapes.py:112]: I wanted to make changes that self-contained as possible. Of course this was before I had to move 1k lines into its own file. I'll move this one now.

**Review by peterbell10 — APPROVED** [review]

File: torch/fx/experimental/_sym_dispatch_mode.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        global SYM_FUNCTION_MODE
+        SYM_FUNCTION_MODE = self.inner
+
+
+def handle_sym_dispatch(func, args, kwargs):
+    global SYM_FUNCTION_MODE
+    mode = sym_function_mode()
+    assert mode
+    SYM_FUNCION_MODE = mode.inner
```
Comment (peterbell10) [inline torch/fx/experimental/_sym_dispatch_mode.py:48]: Typo
```suggestion
    SYM_FUNCTION_MODE = mode.inner
```

File: torch/fx/experimental/_sym_dispatch_mode.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    try:
+        # TODO: properly compute types
+        types: List[Type] = []
+        return mode.__sym_dispatch__(func, types, args, kwargs)
+    finally:
+        SYM_FUNCION_MODE = mode
```
Comment (peterbell10) [inline torch/fx/experimental/_sym_dispatch_mode.py:54]:
```suggestion
        SYM_FUNCTION_MODE = mode
```

**Pull request closed without merging**